expression/collation.go

```diff
@@ -168,131 +192,272 @@ func deriveCoercibilityForColumn(c *Column) Coercibility {
+	for _, arg := range args {
+		if arg.GetType().Charset == ec.Charset {
+			continue
+		}
+
+		if arg.Repertoire() == ASCII {
+			continue
+		}
+
+		if c, ok := arg.(*Constant); ok {
+			str, isNull, err := c.EvalString(ctx, chunk.Row{})
+			if err != nil {
+				return false
+			}
+			if !isNull && !isValidString(str, ec.Charset) {
+				return false
+			}
+		} else {
+			if arg.GetType().Collate != charset.CharsetBin && ec.Charset != charset.CharsetBin && !isUnicodeCollation(ec.Charset) {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+// isValidString check if str can convert to sdtChs charset without data lose.
+func isValidString(str string, dstChs string) bool {
+	switch dstChs {
+	case charset.CharsetASCII:
+		for _, c := range str {
+			if c >= 0x80 {
+				return false
+			}
+		}
+		return true
+	case charset.CharsetLatin1:
+		// for backward compatibility, we do not block SQL like select '啊' = convert('a' using latin1) collate latin1_bin;
+		return true
+	case charset.CharsetUTF8, charset.CharsetUTF8MB4:
+		// string in tidb is actually use utf8mb4 encoding.
+		return true
+	case charset.CharsetBinary:
+		// convert to binary is always safe.
+		return true
+	default:
+		e, _ := charset.Lookup(dstChs)
+		_, err := e.NewEncoder().String(str)
+		return err == nil
+	}
 }
 
 // inferCollation infers collation, charset, coercibility and check the legitimacy.
-func inferCollation(exprs ...Expression) (dstCollation, dstCharset string, coercibility Coercibility, legal bool) {
-	firstExplicitCollation := ""
-	coercibility = CoercibilityIgnorable
-	dstCharset, dstCollation = charset.GetDefaultCharsetAndCollate()
-	for _, arg := range exprs {
-		if arg.Coercibility() == CoercibilityExplicit {
-			if firstExplicitCollation == "" {
-				firstExplicitCollation = arg.GetType().Collate
-				coercibility, dstCollation, dstCharset = CoercibilityExplicit, arg.GetType().Collate, arg.GetType().Charset
-			} else if firstExplicitCollation != arg.GetType().Collate {
-				return "", "", CoercibilityIgnorable, false
+func inferCollation(exprs ...Expression) *ExprCollation {
+	if len(exprs) == 0 {
+		// TODO: see if any function with no arguments could run here.
+		dstCharset, dstCollation := charset.GetDefaultCharsetAndCollate()
+		return &ExprCollation{
+			Coer:      CoercibilityIgnorable,
+			Repe:      UNICODE,
+			Charset:   dstCharset,
+			Collation: dstCollation,
+		}
+	}
+
+	repertoire := exprs[0].Repertoire()
+	coercibility := exprs[0].Coercibility()
+	dstCharset, dstCollation := exprs[0].GetType().Charset, exprs[0].GetType().Collate
+	unknownCS := false
+
+	for _, arg := range exprs[1:] {
+		if dstCollation == charset.CollationBin || arg.GetType().Collate == charset.CollationBin {
+			if coercibility > arg.Coercibility() || (coercibility == arg.Coercibility() && arg.GetType().Collate == charset.CollationBin) {
+				coercibility, dstCharset, dstCollation = arg.Coercibility(), arg.GetType().Charset, arg.GetType().Collate
```

Comment:
Why do we need to do this when `arg.GetType().Collate == charset.CollationBin`  is required? Do we need to add some comments for this logic？Some of the logic in this function is not easy to understand.

Reply:
sure, I will add comments about it.

Reply:
the comment is updated. PTAL